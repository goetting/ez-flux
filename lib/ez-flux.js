'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _minimitter = require('./minimitter');

var _minimitter2 = _interopRequireDefault(_minimitter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var isFn = function isFn(fn) {
  return typeof fn === 'function';
};

var EZFlux = function (_MiniMitter) {
  _inherits(EZFlux, _MiniMitter);

  _createClass(EZFlux, null, [{
    key: 'getEventNames',
    value: function getEventNames(stateName) {
      var actionName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

      return {
        triggered: 'triggered:action.' + stateName + '.' + actionName,
        canceled: 'canceled:action.' + stateName + '.' + actionName,
        change: 'change:state.' + stateName,
        reset: 'RESET:state.' + stateName
      };
    }
  }, {
    key: 'getActionCycle',
    value: function getActionCycle(action, beforeActions, afterActions) {
      var cycle = [action];

      if (beforeActions) cycle.push(beforeActions);
      if (afterActions) cycle.unshift(afterActions);

      return cycle;
    }
  }, {
    key: 'validateScope',
    value: function validateScope(name, _ref) {
      var values = _ref.values,
          actions = _ref.actions,
          afterActions = _ref.afterActions,
          beforeActions = _ref.beforeActions;

      if (!values || (typeof values === 'undefined' ? 'undefined' : _typeof(values)) !== 'object') {
        throw new Error('ezFlux: "' + name + '" - must include a values object');
      }
      if (!actions || Object.keys(actions).find(function (key) {
        return !isFn(actions[key]);
      })) {
        throw new Error('ezFlux: "' + name + '" - actions must include dictionary of functions');
      }
      if (afterActions && !isFn(afterActions)) {
        throw new Error('ezFlux: "' + name + '" - \'afterActions\' must be a function or undefined');
      }
      if (beforeActions && !isFn(beforeActions)) {
        throw new Error('ezFlux: "' + name + '" - \'beforeActions\' must be a function or undefined');
      }
    }
  }]);

  function EZFlux() {
    var stateCfg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, EZFlux);

    var _this = _possibleConstructorReturn(this, (EZFlux.__proto__ || Object.getPrototypeOf(EZFlux)).call(this));

    _this.history = {};
    _this.config = {};
    _this.actions = {};
    _this.defaultState = {};
    _this.state = {};
    _this.plugins = {};

    var scopeNames = Object.keys(stateCfg);
    var initState = options.initialState || {};

    _this.config = options;

    for (var i = scopeNames.length; i--;) {
      var _name = scopeNames[i];
      var scopeConfig = stateCfg[_name];

      _this.constructor.validateScope(_name, scopeConfig);
      _this.addScopeToState(_name, scopeConfig.values, initState[_name]);
      _this.addScopeToEventSystem(_name, scopeConfig);
    }
    Object.freeze(_this.state);

    if (options.plugins) {
      for (var _i = options.plugins.length; _i--;) {
        _this.plug(options.plugins[_i]);
      }delete _this.config.plugins;
    }
    return _this;
  }

  _createClass(EZFlux, [{
    key: 'addScopeToState',
    value: function addScopeToState(name, values) {
      var initValues = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

      var keys = Object.keys(values);

      this.state[name] = {};
      for (var i = keys.length; i--;) {
        var key = keys[i];
        var val = values[key];

        if (!isFn(val)) this.state[name][key] = initValues[key] || val;else Object.defineProperty(this.state[name], key, { enumerable: true, get: val.bind(this) });
      }
      this.defaultState[name] = Object.freeze(this.state[name]);
    }
  }, {
    key: 'addScopeToEventSystem',
    value: function addScopeToEventSystem(scopeName, scopeConfig) {
      var beforeActions = scopeConfig.beforeActions,
          afterActions = scopeConfig.afterActions,
          actions = scopeConfig.actions;

      var actionNames = Object.keys(actions);

      this.actions[scopeName] = {};

      for (var i = actionNames.length; i--;) {
        var _actionName = actionNames[i];
        var eventNames = this.constructor.getEventNames(scopeName, _actionName);
        var action = actions[_actionName];
        var actionCycle = this.constructor.getActionCycle(action, beforeActions, afterActions);
        var listener = this.getActionListener(scopeName, _actionName, actionCycle, eventNames);
        var trigger = this.getActionTrigger(eventNames);

        this.actions[scopeName][_actionName] = trigger;
        this.on(eventNames.triggered, listener);
      }
    }
  }, {
    key: 'getActionTrigger',
    value: function getActionTrigger(eventNames) {
      var _this2 = this;

      return function (payload) {
        return new Promise(function (res, rej) {
          return _this2.fluxEmit(eventNames.triggered, payload, res, rej);
        });
      };
    }
  }, {
    key: 'getActionListener',
    value: function getActionListener(scopeName, actionName, actionCycle, eventNames) {
      var _this3 = this;

      return function (payload, res, rej) {
        var i = actionCycle.length - 1;
        var stateChange = Object.seal(_extends({}, _this3.state[scopeName]));
        var runSeries = function runSeries(actions, cb) {
          var result = actions[i](payload, stateChange, _this3, actionName);
          var validateActionResult = function validateActionResult(actionResult) {
            if (!actionResult || (typeof actionResult === 'undefined' ? 'undefined' : _typeof(actionResult)) !== 'object') {
              cb(false);
              return;
            }
            try {
              Object.assign(stateChange, actionResult);
              i -= 1;
              if (actions[i]) runSeries(actions, cb);else cb(true);
            } catch (err) {
              rej(err);
            }
          };

          if (!result || !result.then) validateActionResult(result);else result.then(validateActionResult).catch(rej);
        };

        runSeries(actionCycle, function (success) {
          if (success) {
            _this3.setStateScope(scopeName, stateChange, eventNames.change);
            res(stateChange);
          } else {
            _this3.fluxEmit(eventNames.canceled);
            res();
          }
        });
      };
    }
  }, {
    key: 'setStateScope',
    value: function setStateScope(name, newState, eventName) {
      this.state = _extends({}, this.state);
      this.state[name] = _extends({}, newState);
      Object.freeze(this.state);
      Object.freeze(this.state[name]);
      this.fluxEmit(eventName);
    }
  }, {
    key: 'fluxEmit',
    value: function fluxEmit() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
      var payload = arguments[1];
      var res = arguments[2];
      var rej = arguments[3];

      this.emit(name, payload, res, rej);

      if (this.config.recordHistory) {
        var _time = Date.now();
        var _state = {};
        var scopes = Object.keys(this.state);

        for (var i = scopes.length; i--;) {
          _state[scopes[i]] = _extends({}, this.state[scopes[i]]);
        }this.history[_time] = { time: _time, name: name, state: _state, payload: payload };
      }
      if (this.config.onFluxEmit) {
        this.config.onFluxEmit(name, payload, this);
      }
      return this;
    }
  }, {
    key: 'resetStateScope',
    value: function resetStateScope(name) {
      if (!this.defaultState[name]) throw new Error('ezFlux.reset: ' + name + ' not found on state');
      this.setStateScope(name, this.defaultState[name], this.constructor.getEventNames(name).reset);
    }
  }, {
    key: 'resetState',
    value: function resetState() {
      var names = Object.keys(this.defaultState);

      for (var i = names.length; i--;) {
        this.resetStateScope(names[i]);
      }
    }
  }, {
    key: 'plug',
    value: function plug(fn) {
      if (!isFn(fn)) throw new Error('ezFlux: plugin must be a function');
      Object.assign(this.plugins, fn.apply(this));
    }
  }]);

  return EZFlux;
}(_minimitter2.default);

exports.default = EZFlux;