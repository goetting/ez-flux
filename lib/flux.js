'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _eventemitter = require('eventemitter3');

var _eventemitter2 = _interopRequireDefault(_eventemitter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var colorMap = {
  error: 'red',
  action: 'cyan',
  remote: 'magenta',
  tracking: 'gray',
  change: 'green',
  removeListener: 'gray'
};

var Flux = function () {
  function Flux() {
    var _this = this;

    var stateConfig = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var cfg = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, Flux);

    this.bus = new _eventemitter2.default();
    this.state = {};
    this.stateGetters = { get: function get() {
        return Object.assign({}, _this.state);
      } };
    this.actionTriggers = {};
    this.updateBuffer = [];
    this.updateTimeout = null;

    this.cfg = cfg;

    this.bus.setMaxListeners(cfg.maxListeners);
    this.bus.history = {};

    Object.keys(stateConfig).forEach(function (name) {
      return _this.addStateScope(name, stateConfig[name]);
    });
  }

  /**
                              State Configuration
  */

  _createClass(Flux, [{
    key: 'addStateScope',
    value: function addStateScope(name, _ref) {
      var _this2 = this;

      var state = _ref.state,
          actions = _ref.actions;

      if (!state || (typeof state === 'undefined' ? 'undefined' : _typeof(state)) !== 'object') {
        if (this.cfg.debug) console.error('state "' + name + '" must include a state object'); // eslint-disable-line no-console
        return;
      }
      if (!actions || Object.keys(actions).find(function (key) {
        return typeof actions[key] !== 'function';
      })) {
        if (this.cfg.debug) console.error('state "' + name + '" actions must include dictionary of functions'); // eslint-disable-line no-console
        return;
      }
      this.state[name] = Object.assign({}, state);
      this.stateGetters[name] = { get: function get() {
          return Object.assign({}, _this2.state[name]);
        } };

      Object.keys(actions).forEach(function (actionName) {
        var eventName = _this2.createActionTrigger(name, actionName);
        var action = actions[actionName];
        var setState = function setState(stateChange) {
          if (!stateChange || (typeof stateChange === 'undefined' ? 'undefined' : _typeof(stateChange)) !== 'object') {
            if (_this2.cfg.debug) console.error(name + '.' + actionName + ': setState argument must be Object'); // eslint-disable-line no-console
            return;
          }
          Object.assign(_this2.state[name], stateChange);
          _this2.queueUpdate(name);
        };

        _this2.bus.on(eventName, function (data) {
          action(data, _this2.state[name].get(), setState);
        });
      });
    }
  }, {
    key: 'createActionTrigger',
    value: function createActionTrigger(stateName, actionName) {
      var _this3 = this;

      var eventName = 'action.' + stateName + '.' + actionName;

      if (!this.actionTriggers[stateName]) this.actionTriggers[stateName] = {};
      this.actionTriggers[stateName][actionName] = function (data) {
        return _this3.bus.emit(eventName, data || {});
      };

      return eventName;
    }

    /**
                                Event Handling
    */

  }, {
    key: 'flushUpdates',
    value: function flushUpdates() {
      var key = this.updateBuffer.shift();
      if (!key) return;

      this.emit('state.change.' + key, Object.assign(this.state[key]));
      this.flushUpdates();
    }
  }, {
    key: 'queueUpdate',
    value: function queueUpdate(key) {
      if (this.updateBuffer.indexOf(key) === -1) this.updateBuffer.push(key);
      if (this.cfg.runsInBrowser || this.cfg.throttleUpdates) {
        this.flushUpdates();
        return;
      }

      if (!(this.cfg.runsInBrowser && window.requestAnimationFrame)) {
        clearTimeout(this.updateTimeout);
        this.updateTimeout = setTimeout(this.flushUpdates, 60 / 1000);
      } else {
        window.cancelAnimationFrame(this.updateTimeout);
        this.updateTimeout = window.requestAnimationFrame(this.flushUpdates);
      }
    }
  }, {
    key: 'emit',
    value: function emit() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

      for (var _len = arguments.length, args = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
        args[_key - 1] = arguments[_key];
      }

      if (this.cfg.debug) {
        var data = [].concat(_toConsumableArray(args));
        var time = new Date().getTime();
        var unsub = name === 'removeListener' ? data.shift() : '';
        var msg = 'ezFlux | ' + time + ' ' + name + ' ' + unsub;
        var color = colorMap[name.split('.')[0]] || 'gray';

        this.bus.history[time + ' ' + msg] = { time: time, msg: msg, data: data };

        if (this.cfg.runsInBrowser) {
          var _console;

          if (!this.cfg.logAppState) console.log('%c' + msg, 'color:' + color); // eslint-disable-line no-console
          else (_console = console).log.apply(_console, ['%c' + msg, 'color:' + color, '\n\n'].concat(_toConsumableArray(data), ['\n\n'])); // eslint-disable-line no-console
        }
      }
      this.bus.emit.apply(this.bus, [name].concat(_toConsumableArray(args)));
    }
  }]);

  return Flux;
}();

exports.default = Flux;