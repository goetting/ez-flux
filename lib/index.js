'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var colorMap = { RESET: 'red', trigger: 'cyan', change: 'green' };
var isFn = function isFn(fn) {
  return typeof fn === 'function';
};

var EZFlux = function () {
  _createClass(EZFlux, null, [{
    key: 'getEventNames',
    value: function getEventNames(stateName) {
      var actionName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

      return {
        triggered: 'triggered:action.' + stateName + '.' + actionName,
        canceled: 'canceled:action.' + stateName + '.' + actionName,
        change: 'change:state.' + stateName,
        reset: 'RESET:state.' + stateName
      };
    }
  }, {
    key: 'getActionCycle',
    value: function getActionCycle(action, beforeActions, afterActions) {
      var cycle = [action];

      if (beforeActions) cycle.push(beforeActions);
      if (afterActions) cycle.unshift(afterActions);

      return cycle;
    }
  }, {
    key: 'validateScope',
    value: function validateScope(name, _ref) {
      var values = _ref.values,
          actions = _ref.actions,
          afterActions = _ref.afterActions,
          beforeActions = _ref.beforeActions;

      if (!values || (typeof values === 'undefined' ? 'undefined' : _typeof(values)) !== 'object') {
        throw new Error('ezFlux: "' + name + '" - must include a values object');
      }
      if (!actions || Object.keys(actions).find(function (key) {
        return !isFn(actions[key]);
      })) {
        throw new Error('ezFlux: "' + name + '" - actions must include dictionary of functions');
      }
      if (afterActions && !isFn(afterActions)) {
        throw new Error('ezFlux: "' + name + '" - \'afterActions\' must be a function or undefined');
      }
      if (beforeActions && !isFn(beforeActions)) {
        throw new Error('ezFlux: "' + name + '" - \'beforeActions\' must be a function or undefined');
      }
    }
  }]);

  function EZFlux() {
    var stateCfg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    _classCallCheck(this, EZFlux);

    this.history = {};
    this.config = {};
    this.runsInBrowser = typeof window !== 'undefined' && !!window.requestAnimationFrame;
    this.actions = {};
    this.emissionTimeout = null;
    this.defaultState = {};
    this.state = {};
    this.events = {};
    this.removeListener = this.off;
    this.plugins = {};

    var scopeNames = Object.keys(stateCfg);
    var initState = options.initialState || {};

    this.config = options;

    for (var i = scopeNames.length; i--;) {
      var _name = scopeNames[i];
      var scopeConfig = stateCfg[_name];

      this.constructor.validateScope(_name, scopeConfig);

      this.state[_name] = _extends({}, scopeConfig.values);
      this.defaultState[_name] = _extends({}, scopeConfig.values);
      if (initState[_name]) Object.assign(this.state[_name], initState[_name]);
      Object.freeze(this.state[_name]);

      this.addScopeToEventSystem(_name, scopeConfig);
    }
    Object.freeze(this.state);

    if (options.plugins) {
      for (var _i = options.plugins.length; _i--;) {
        this.plug(options.plugins[_i]);
      }delete this.config.plugins;
    }
  }

  _createClass(EZFlux, [{
    key: 'addScopeToEventSystem',
    value: function addScopeToEventSystem(scopeName, scopeConfig) {
      var _this = this;

      var actionNames = Object.keys(scopeConfig.actions);
      var beforeActions = scopeConfig.beforeActions,
          afterActions = scopeConfig.afterActions,
          actions = scopeConfig.actions;


      this.actions[scopeName] = {};

      var _loop = function _loop(i) {
        var actionName = actionNames[i];
        var eventNames = _this.constructor.getEventNames(scopeName, actionName);
        var action = actions[actionName];
        var actionCycle = _this.constructor.getActionCycle(action, beforeActions, afterActions);
        var listener = _this.getActionListener(scopeName, actionName, actionCycle, eventNames);
        var trigger = function trigger(payload) {
          return new Promise(function (res) {
            return _this.emit(eventNames.triggered, payload, res);
          });
        }; // eslint-disable-line no-loop-func

        _this.actions[scopeName][actionName] = trigger;
        _this.on(eventNames.triggered, listener);
      };

      for (var i = actionNames.length; i--;) {
        _loop(i);
      }
    }
  }, {
    key: 'getActionListener',
    value: function getActionListener(scopeName, actionName, actionCycle, eventNames) {
      var _this2 = this;

      return function (payload, res) {
        var i = actionCycle.length - 1;
        var stateChange = Object.seal(_extends({}, _this2.state[scopeName]));
        var runSeries = function runSeries(actions, cb) {
          var result = actions[i](payload, stateChange, _this2, actionName);
          var validateActionResult = function validateActionResult(actionResult) {
            if (!actionResult || (typeof actionResult === 'undefined' ? 'undefined' : _typeof(actionResult)) !== 'object') {
              cb(false);
              return;
            }
            Object.assign(stateChange, actionResult);
            i -= 1;
            if (actions[i]) runSeries(actions, cb);else cb(true);
          };

          if (!result || !result.then) validateActionResult(result);else result.then(validateActionResult);
        };

        runSeries(actionCycle, function (success) {
          if (success) {
            _this2.setStateScope(scopeName, stateChange);
            _this2.emit(eventNames.change);
          } else {
            _this2.emit(eventNames.canceled);
          }
          res();
        });
      };
    }
  }, {
    key: 'setStateScope',
    value: function setStateScope(name, newState) {
      this.state = _extends({}, this.state);
      this.state[name] = _extends({}, newState);
      Object.freeze(this.state);
      Object.freeze(this.state[name]);
    }
  }, {
    key: 'logEmission',
    value: function logEmission() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
      var payload = arguments[1];

      if (this.config.recordHistory) {
        var _time = Date.now();
        var _state = {};
        var scopes = Object.keys(this.state);

        for (var i = scopes.length; i--;) {
          _state[scopes[i]] = _extends({}, this.state[scopes[i]]);
        }this.history[_time] = { time: _time, name: name, state: _state, payload: payload };
      }

      if (this.config.console && console[this.config.console]) {
        // eslint-disable-line no-console
        var logger = console[this.config.console]; // eslint-disable-line no-console
        var msg = 'ezFlux | ' + name;
        var color = colorMap[name.split(':')[0]] || 'gray';
        logger.apply(undefined, _toConsumableArray(this.runsInBrowser ? ['%c' + msg, 'color:' + color] : [msg]));
      }
    }
  }, {
    key: 'emit',
    value: function emit() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
      var payload = arguments[1];
      var resolver = arguments[2];

      this.logEmission(name, payload);
      if (!this.events[name]) return;
      for (var i = this.events[name].length; i--;) {
        this.events[name][i](payload, resolver);
      }
    }
  }, {
    key: 'on',
    value: function on(name, fn) {
      if (!this.events[name]) this.events[name] = [fn];else this.events[name].push(fn);
    }
  }, {
    key: 'once',
    value: function once(name, fn) {
      var _this3 = this;

      this.on(name, function () {
        fn.apply(undefined, arguments);
        _this3.off(name, fn);
      });
    }
  }, {
    key: 'off',
    value: function off(name, fn) {
      if (!this.events[name]) return;
      var i = this.events[name].findIndex(function (handler) {
        return handler === fn;
      });

      if (i > -1) this.events[name].splice(i, 1);
    }
  }, {
    key: 'resetStateScope',
    value: function resetStateScope(name) {
      if (!this.defaultState[name]) throw new Error('ezFlux.reset: ' + name + ' not found on state');
      this.setStateScope(name, this.defaultState[name]);
      this.emit(this.constructor.getEventNames(name).reset);
    }
  }, {
    key: 'resetState',
    value: function resetState() {
      var names = Object.keys(this.defaultState);

      for (var i = names.length; i--;) {
        this.resetStateScope(names[i]);
      }
    }
  }, {
    key: 'plug',
    value: function plug(fn) {
      if (!isFn(fn)) throw new Error('ezFlux: plugin must be a function');
      Object.assign(this.plugins, fn.apply(this));
    }
  }]);

  return EZFlux;
}();

exports.default = EZFlux;