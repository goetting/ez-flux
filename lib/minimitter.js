'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var MiniMitter = function () {
  function MiniMitter() {
    _classCallCheck(this, MiniMitter);

    this.events = {};
    this.addListener = this.on;
    this.removeListener = this.off;
  }

  _createClass(MiniMitter, [{
    key: 'emit',
    value: function emit() {
      var name = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

      if (this.events[name]) {
        for (var _len = arguments.length, payload = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
          payload[_key - 1] = arguments[_key];
        }

        for (var i = this.events[name].length; i--;) {
          var fn = this.events[name][i];
          fn.apply(undefined, _toConsumableArray(payload));
          if (fn.once) this.off(name, fn);
        }
      }
      return this;
    }
  }, {
    key: 'on',
    value: function on(name, fn) {
      this.emit('newListener', name, fn);
      if (!this.events[name]) this.events[name] = [fn];else this.events[name].push(fn);
      return this;
    }
  }, {
    key: 'once',
    value: function once(name, fn) {
      fn.once = true; // eslint-disable-line no-param-reassign
      this.on(name, fn);
      return this;
    }
  }, {
    key: 'off',
    value: function off(name, fn) {
      if (this.events[name]) {
        var i = this.events[name].indexOf(fn);

        if (i > -1) this.events[name].splice(i, 1);
        this.emit('removeListener', name, fn);
      }
      return this;
    }
  }]);

  return MiniMitter;
}();

exports.default = MiniMitter;