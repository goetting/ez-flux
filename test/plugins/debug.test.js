/* eslint-disable no-unused-expressions, no-use-before-define */
import { createStore, plugins } from '../../lib/index';
import createDebugger from '../../lib/plugins/debug';

describe('debug plugin', () => {
  afterEach(() => { plugins.length = 0; });

  describe('createDebugger', () => {
    it('should throw if first arg is not a function', noCB);
    it('should create a pluggable function', returnsFn);
  });
  describe('debug', () => {
    it('should throw an error if a given store has no name', noName);
    it('should not wrap native store methods', leaveNative);
    it('should wrap optional methods', logMethod);
    it('should subscribe to and log a change event', logChange);
    it('should log child events differently', logChild);
  });
});


function noCB() {
  let err = null;
  try {
    createDebugger();
  } catch (e) {
    err = e;
  }
  expect(err).toBeTruthy();
}

function returnsFn() {
  expect(typeof createDebugger(() => {})).toBe('function');
}

function noName() {
  let err = null;
  try {
    plugins.push(createDebugger(() => {}));
    const store = createStore({});

    store.$emit('change');
  } catch (e) {
    err = e;
  }
  expect(err).toBeTruthy();
}

function leaveNative() {
  let cbCount = 0;

  plugins.push(createDebugger((payload) => {
    if (payload.eventType === 'method called') cbCount++;
  }));

  const store = createStore({ name: 'store' });

  store.$reset();
  expect(cbCount).toBe(0);
}

function logMethod() {
  let cbCount = 0;

  plugins.push(createDebugger((payload) => {
    cbCount++;
    expect(payload).toMatchSnapshot();
  }));

  const store = createStore({ name: 'store', methods: { someMethod: () => {} } });

  store.someMethod();
  expect(cbCount).toBe(1);
}

function logChange() {
  let cbCount = 0;

  plugins.push(createDebugger((payload) => {
    cbCount++;
    expect(payload).toMatchSnapshot();
  }));

  const store = createStore({ name: 'store', state: { foo: 'bar' } });

  store.$assign({ foo: 'bar' });
  expect(cbCount).toBe(1);
}

function logChild() {
  let cbCount = 0;

  plugins.push(createDebugger((payload) => {
    cbCount++;
    expect(payload).toMatchSnapshot();
  }));

  const child = createStore({ name: 'child', state: { foo: 'bar' } });
  createStore({ name: 'parent', state: { bar: 'baz' }, children: { child } });

  child.$assign({ foo: 'baz' });
  expect(cbCount).toBe(2);
}

