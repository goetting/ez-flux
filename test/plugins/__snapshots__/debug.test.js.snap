// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`debug plugin debug should log child events differently 1`] = `
Object {
  "childName": "child",
  "eventType": "child change",
  "state": Object {
    "bar": "baz",
    "child": Object {
      "foo": "baz",
    },
  },
  "store": Object {
    "$assign": [Function],
    "$copy": [Function],
    "$emit": [Function],
    "$entries": [Function],
    "$events": Object {
      "change": Array [
        [Function],
      ],
    },
    "$keys": [Function],
    "$off": [Function],
    "$on": [Function],
    "$once": [Function],
    "$reset": [Function],
    "$values": [Function],
    "bar": "baz",
    "child": Object {
      "$assign": [Function],
      "$copy": [Function],
      "$emit": [Function],
      "$entries": [Function],
      "$events": Object {
        "change": Array [
          [Function],
          [Function],
        ],
      },
      "$keys": [Function],
      "$off": [Function],
      "$on": [Function],
      "$once": [Function],
      "$reset": [Function],
      "$values": [Function],
      "foo": "baz",
    },
  },
  "storeName": "parent",
}
`;

exports[`debug plugin debug should log child events differently 2`] = `
Object {
  "childName": undefined,
  "eventType": "state change",
  "state": Object {
    "foo": "baz",
  },
  "store": Object {
    "$assign": [Function],
    "$copy": [Function],
    "$emit": [Function],
    "$entries": [Function],
    "$events": Object {
      "change": Array [
        [Function],
        [Function],
      ],
    },
    "$keys": [Function],
    "$off": [Function],
    "$on": [Function],
    "$once": [Function],
    "$reset": [Function],
    "$values": [Function],
    "foo": "baz",
  },
  "storeName": "child",
}
`;

exports[`debug plugin debug should subscribe to and log a change event 1`] = `
Object {
  "childName": undefined,
  "eventType": "state change",
  "state": Object {
    "foo": "bar",
  },
  "store": Object {
    "$assign": [Function],
    "$copy": [Function],
    "$emit": [Function],
    "$entries": [Function],
    "$events": Object {
      "change": Array [
        [Function],
      ],
    },
    "$keys": [Function],
    "$off": [Function],
    "$on": [Function],
    "$once": [Function],
    "$reset": [Function],
    "$values": [Function],
    "foo": "bar",
  },
  "storeName": "store",
}
`;

exports[`debug plugin debug should wrap optional methods 1`] = `
Object {
  "eventType": "method called",
  "methodName": "someMethod",
  "state": Object {},
  "store": Object {
    "$assign": [Function],
    "$copy": [Function],
    "$emit": [Function],
    "$entries": [Function],
    "$events": Object {
      "change": Array [
        [Function],
      ],
    },
    "$keys": [Function],
    "$off": [Function],
    "$on": [Function],
    "$once": [Function],
    "$reset": [Function],
    "$values": [Function],
    "someMethod": [Function],
  },
  "storeName": "store",
}
`;
